crates/sui-core/src/authority.rs

```diff
@@ -2321,19 +2299,33 @@ impl AuthorityState {
     }
 
     /// Make an information response for a transaction
-    pub async fn make_transaction_info(
+    pub fn make_transaction_info(
         &self,
         transaction_digest: &TransactionDigest,
         epoch_store: &Arc<AuthorityPerEpochStore>,
-    ) -> Result<VerifiedTransactionInfoResponse, SuiError> {
-        Ok(VerifiedTransactionInfoResponse {
-            signed_transaction: epoch_store.get_signed_transaction(transaction_digest)?,
-            certified_transaction: self
+    ) -> Result<Option<VerifiedTransactionInfoResponse>, SuiError> {
+        if let Some(effects) =
+            self.get_signed_effects_and_maybe_resign(epoch_store.epoch(), transaction_digest)?
+        {
+            if let Some(cert) = self
                 .database
-                .get_certified_transaction(transaction_digest)?,
-            signed_effects: self
-                .get_signed_effects_and_maybe_resign(epoch_store.epoch(), transaction_digest)?,
-        })
+                .get_certified_transaction(transaction_digest)?
+            {
+                return Ok(Some(VerifiedTransactionInfoResponse::Executed(
+                    cert, effects,
+                )));
+            }
+            // The read of effects and read of cert are not atomic. It's possible that we reverted
+            // the transaction (during epoch change) in between the above two reads, and we end up
+            // having effects but not certs. In this case, we just fall through.
```

Comment:
if this is the only case where this can happen (i believe it is) can we log a warning and return Ok(None)? I think its easier on clients if we don't expose them to this very rare edge case.